src/msix/pack/AppxBundleWriter.cpp

```diff
@@ -0,0 +1,280 @@
+
+                if (flatBundle)
+                {
+                    ThrowHrIfFailed(AddPackageReference(utf8_to_wstring(file.second).c_str(), stream.Get(), 
+                        false));
+                }
+            }
+        }
+
+        //Process external packages passed as input created from mapping file
+        /*if (externalPackagesList != nullptr)
+        {
+
+        }*/
+
+        failState.release();
+    }
+
+    // IAppxBundleWriter
+    HRESULT STDMETHODCALLTYPE AppxBundleWriter::AddPayloadPackage(LPCWSTR fileName, IStream* packageStream) noexcept try
+    {
+        // TODO: implement
+        NOTIMPLEMENTED;
+    } CATCH_RETURN();
+
+    HRESULT STDMETHODCALLTYPE AppxBundleWriter::Close() noexcept try
+    {
+        std::string targetXmlNamespace;
+        //validate namespace according to input and assign namespace
+        //m_bundleManifestWriter.StartBundleManifest(targetXmlNamespace, wstring_to_utf8(this->mainPackageName),
+          //  wstring_to_utf8(this->mainPackagePublisher), this->bundleVersion);
+
+        for(std::size_t i = 0; i < this->m_bundleWriterHelper.GetPayloadPackages().size(); i++) 
+        {
+            //Call AddPackage for each package
+            //this->payloadPackages[i].architecture
+        }
+
+        //Do the same loop for this->OptionalBundles
+        
+        //EndBundleManifest -> Ends Packages Element
+        m_bundleManifestWriter.EndPackagesElement();
+
+        // Add AppxBundleManifest.xml to blockmap and zip
+        m_bundleManifestWriter.Close();
+        auto bundleManifestStream = m_bundleManifestWriter.GetStream();
+        auto bundleManifestContentType = ContentType::GetBundlePayloadFileContentType(APPX_BUNDLE_FOOTPRINT_FILE_TYPE_MANIFEST);
+        AddFileToPackage(APPXBUNDLEMANIFEST_XML, bundleManifestStream.Get(), true, true, bundleManifestContentType.c_str());
+
+        CloseInternal();
+        return static_cast<HRESULT>(Error::OK);
+    } CATCH_RETURN();
+
+    // IAppxBundleWriter4
+    HRESULT STDMETHODCALLTYPE AppxBundleWriter::AddPackageReference(LPCWSTR fileName, 
+        IStream* inputStream, BOOL isDefaultApplicablePackage) noexcept try
+    {        
+        HRESULT hr = this->AddPackageReferenceInternal(fileName, inputStream, !!isDefaultApplicablePackage);
+
+        if (FAILED(hr))
+        {
+            //fail
+        }
+
+        return hr;
+
+    } CATCH_RETURN();
+
+    HRESULT AppxBundleWriter::AddPackageReferenceInternal(
+        _In_ LPCWSTR fileName,
+        _In_ IStream* packageStream,
+        _In_ bool isDefaultApplicablePackage)
+    {
+        auto appxFactory = m_factory.As<IAppxFactory>();
+
+        ComPtr<IAppxPackageReader> reader;
+        ThrowHrIfFailed(appxFactory->CreatePackageReader(packageStream, &reader));
+
+        UINT64 packageStreamSize = 0;
+        ThrowHrIfFailed(this->m_bundleWriterHelper.GetStreamSize(packageStream, &packageStreamSize));

```

Comment:
>ThrowHrIfFailed [](start = 8, length = 15)

when you use this, the functions need to CATCH_RETURN()

Reply:
Updated to return void for all internal methods except methods implementing an interface(which returns HRESULT and uses CATCH_RETURN()